### PR TITLE
Cleanup method

### PR DESCRIPTION
**Pull request by nwlandry**
* Added a cleanup method that removes artifacts specified by the user: multi-edges, singletons, isolates. It also can convert all labels to consecutive integers.
* Modified the `duplicates()` method to not include the first instance of the node/edge in the list of duplicates, so that when you call `H.remove_edges_from(H.edges.duplicates())` it only removes the redundant edges, not all of the edges. In preparation of only supporting non-loopy edges, the `duplicates()` method has been simplified to not differentiate between duplicate nodes in an edge.

### CODE REVIEW

File: xgi/classes/hypergraph.py [hunk 1]
```diff
@@ -964,3 +964,47 @@ def dual(self):
         dual._hypergraph = deepcopy(self._hypergraph)
 
         return dual
+
+    def cleanup(
+        self,
+        remove_isolates=True,
+        remove_singletons=True,
+        remove_multiedges=True,
```
Comment (leotrs) [inline xgi/classes/hypergraph.py:972]:
```suggestion
        isolates=True,
        singletons=True,
        multiedges=True,
```

File: xgi/classes/hypergraph.py [hunk 2]
```diff
@@ -964,3 +964,47 @@ def dual(self):
...
+
+        Parameters
+        ----------
+        remove_isolates : bool, optional
+            Whether to remove isolated nodes, by default True
+        remove_singletons : bool, optional
+            Whether to remove singleton edges, by default True
+        remove_multiedges : bool, optional
+            Whether to remove duplicate edges, by default True
+        reindex : bool, optional
```
Comment (leotrs) [inline xgi/classes/hypergraph.py:985]: the name of the function contains the word `label` so to keep it consistent, this parameter should be called `relabel` too

File: xgi/classes/reportviews.py [hunk 3]
```diff
@@ -375,7 +375,8 @@ def duplicates(self):
         Notes
         -----
         The IDs returned are in an arbitrary order, that is duplicates are not
-        guaranteed to be consecutive.
+        guaranteed to be consecutive. For IDs with the same bipartite neighbors,
+        only the first ID added is not a duplicate.
```
Comment (leotrs) [inline xgi/classes/reportviews.py:379]:
```suggestion
        guaranteed to be consecutive.
        If two IDs have the same bipartite neighbors, only one of them is returned.
```

Comment (leotrs) [reply xgi/classes/reportviews.py:379]: I wonder if we should add an option to return all of them, not only those ones that are actually repeated (and ignoring the first one). In a way, `duplicates` is similar to a `groupby` operation (it's grouping IDs that have the same bipartite neighbors) and `groupby` operations always return all of the elements involved, they never ignore the first one.

Comment (leotrs) [reply xgi/classes/reportviews.py:379]: Or am I overthinking things again? :sweat_smile:

File: xgi/convert.py [hunk 4]
```diff
@@ -88,7 +88,9 @@ def convert_to_hypergraph(data, create_using=None):
         from_incidence_matrix(data, create_using)
 
     else:
-        raise XGIError("Not able to convert to a hypergraph because the input data was not a supported type!")
+        raise XGIError(
+            "Not able to convert to a hypergraph because the input data was not a supported type!"
```
Comment (leotrs) [inline xgi/convert.py:92]:
```suggestion
            "Input data has unsupported type."
```

File: xgi/classes/hypergraph.py [hunk 5]
```diff
@@ -964,3 +964,47 @@ def dual(self):
...
+            ee = temp.edges
+
+            self.remove_nodes_from(list(self._node))
+            self.add_nodes_from((n, deepcopy(attr)) for n, attr in nn.items())
+            
+            self.remove_edges_from(list(self._edge))
+            self.add_edges_from(
+            (e, id, deepcopy(temp.edges[id]))
+            for id, e in ee.members(dtype=dict).items()
+        )
```
Comment (leotrs) [inline xgi/classes/hypergraph.py:1009]: Make sure to run black on this.

**Review by leotrs — CHANGES_REQUESTED** [review]
Thanks for this! BTW, does this depend on #162?

Also, should we add a parameter `in_place`? If `in_place` is False, then a new hypergraph is returned and the original one is left intact. This might simplify some of the code inside the `if reindex:` branch.

File: xgi/classes/hypergraph.py (hunk 1 above)
Comment (nwlandry) [reply xgi/classes/hypergraph.py:971]: Done.

File: xgi/classes/hypergraph.py (hunk 2 above)
Comment (nwlandry) [reply xgi/classes/hypergraph.py:984]: Done.

File: xgi/classes/hypergraph.py (hunk 5 above)
Comment (nwlandry) [reply xgi/classes/hypergraph.py:1008]: Done.

File: xgi/classes/reportviews.py (hunk 3 above)
Comment (nwlandry) [reply xgi/classes/reportviews.py:384]: Added as issue [#163](https://github.com/ComplexGroupInteractions/xgi/issues/163).

**Comment by leotrs** [comment]
Pro tip: next time try rebasing your PR rather than merging main into it bc it makes it hard to review. The current diff shows both changes from this PR as well as the other one. If rebasing does not work/is scary, you can also just close it and start a new PR.

**Review by leotrs — APPROVED** [review]

**Pull request merged**